bittensor/utils/registration.py

```diff
@@ -95,5 +110,1003 @@ def __getattr__(self, name):
+                subtensor=subtensor,
+                netuid=netuid,
+                hotkey_bytes=hotkey_bytes,
+                curr_diff=curr_diff,
+                curr_block=curr_block,
+                curr_block_num=curr_block_num,
+                old_block_number=old_block_number,
+                curr_stats=curr_stats,
+                update_curr_block=_update_curr_block,
+                check_block=check_block,
+                solvers=solvers,
+            )
+
+            num_time = 0
+            # Get times for each solver
+            for finished_queue in finished_queues:
+                try:
+                    proc_num = finished_queue.get(timeout=0.1)
+                    num_time += 1
+
+                except Empty:
+                    continue
+
+            time_now = time.time()  # get current time
+            time_since_last = time_now - time_last  # get time since last work block(s)
+            if num_time > 0 and time_since_last > 0.0:
+                # create EWMA of the hash_rate to make measure more robust
+
+                hash_rate_ = (num_time * tpb * update_interval) / time_since_last
+                hash_rates.append(hash_rate_)
+                hash_rates.pop(0)  # remove the 0th data point
+                curr_stats.hash_rate = sum(
+                    [hash_rates[i] * weights[i] for i in range(n_samples)]
+                ) / (sum(weights))
+
+                # update time last to now
+                time_last = time_now
+
+                curr_stats.time_average = (
+                    curr_stats.time_average * curr_stats.rounds_total
+                    + curr_stats.time_spent
+                ) / (curr_stats.rounds_total + num_time)
+                curr_stats.rounds_total += num_time
+
+            # Update stats
+            curr_stats.time_spent = time_since_last
+            new_time_spent_total = time_now - start_time_perpetual
+            curr_stats.hash_rate_perpetual = (
+                curr_stats.rounds_total * (tpb * update_interval)
+            ) / new_time_spent_total
+            curr_stats.time_spent_total = new_time_spent_total
+
+            # Update the logger
+            logger.update(curr_stats, verbose=log_verbose)
+
+        # exited while, found_solution contains the nonce or wallet is registered
+
+        stopEvent.set()  # stop all other processes
+        logger.stop()
+
+        # terminate and wait for all solvers to exit
+        _terminate_workers_and_wait_for_exit(solvers)
+
+        return solution
+
+
+def _terminate_workers_and_wait_for_exit(
+    workers: list[Union[multiprocessing.Process, multiprocessing.queues.Queue]],
+) -> None:
+    """Terminate worker processes and wait for their exit."""
+    for worker in workers:
+        if isinstance(worker, multiprocessing.queues.Queue):
+            worker.join_thread()
+        else:
+            worker.terminate()
+            worker.join()
+        worker.close()
```

Comment:
```suggestion
def _terminate_workers_and_wait_for_exit(
    workers: list[Union[multiprocessing.Process, multiprocessing.queues.Queue]],
) -> None:
    for worker in workers:
        if isinstance(worker, Queue_Type):
            worker.join_thread()
        else:
            worker.join()
        worker.close()
            try:
                worker.join(3.0)
            except subprocess.TimeoutExpired:
                worker.terminate()
        try:
            worker.close()
        except ValueError:
            worker.terminate()
```

We ran into a few edge cases in btcli, where this was a good fix. See: https://github.com/opentensor/btcli/pull/189

Of course you will also need to `import subprocess`

Reply:
Nice. Applyed